### ISSUE TITLE
Contract tests fail on nested out-of-order collections when insertionOrder=false

### ISSUE THREAD

**Issue by osdrv** [aws-cloudformation/cloudformation-cli#813]
Hi folks,

This ticket is a follow-up on https://github.com/aws-cloudformation/cloudformation-cli/issues/811, which solves the problem for flat collections. It seems the problem persists for nested collections with no insertion order.

Assume a schema sample:
```json
{
  "properties": {
    "FooArr": {
      "type": "array",
      "insertionOrder": false,
      "items": {
        "$ref": "#/definitions/Foo"
      }
    }
  },
  "definitions": {
    "FooArr": {
      "type": "object",
      "properties": {
        "FooProps": {
          "type": "array",
          "insertionOrder": false,
          "items": {
            "type": "string"
          }
        }
      }
    }
  }
}
```
In this case, if the order of elements in FooArr.FooProps differes, md5 hashes of object traversals would differ and contract tests would fail in https://github.com/aws-cloudformation/cloudformation-cli/blob/653024cfaab7ecfb8ba7c70212f2fecbabb4b095/src/rpdk/core/jsonutils/utils.py#L22

In other words, I believe `encoded = json.dumps(item, sort_keys=True).encode()` is not recursive and hence is not tolerating a disorder in nested collections.

**Comment by ammokhov** [comment]
oh it's a good catch, the `item_hash` should be changed as:

```python
def item_hash(item):
    dhash = hashlib.md5()  # nosec
    if isinstance(item, dict):
        item = {k: item_hash(v) for k, v in item.items()}
    if isinstance(item, list):
        item = [item_hash(i) for i in item].sort()
    encoded = json.dumps(item, sort_keys=True).encode()
    dhash.update(encoded)
    return dhash.hexdigest()

```

**Comment by ammokhov** [comment]
@osdrv would you provide your input/output so it could be used as a test case?
